File: ga/4omini_result_2024_7/result/multi_thread_basic_mini/14_multi_thread_basic.c
use std::sync::{Arc, Mutex};
use std::thread;

const INITIAL_BALANCE: i32 = 1_000_000; // Global shared variable representing account balance
const WITHDRAW_AMOUNT: i32 = 500_000;    // Amount to withdraw

fn withdraw(account_balance: Arc<Mutex<i32>>, amount: i32) {
    for _ in 0..amount {
        // Lock the mutex to access the shared account balance
        let mut balance = account_balance.lock().unwrap();
        *balance -= 1; // Decrement the balance by 1
    }
}

fn main() {
    // Create an Arc (Atomic Reference Counted) Mutex to share the account balance across threads
    let account_balance = Arc::new(Mutex::new(INITIAL_BALANCE));
    
    // Clone the Arc to pass to the threads
    let account_balance1 = Arc::clone(&account_balance);
    let account_balance2 = Arc::clone(&account_balance);
    
    // Create the withdrawal threads
    let t1 = thread::spawn(move || withdraw(account_balance1, WITHDRAW_AMOUNT));
    let t2 = thread::spawn(move || withdraw(account_balance2, WITHDRAW_AMOUNT));

    // Wait for the threads to finish
    t1.join().unwrap();
    t2.join().unwrap();

    // Print the final account balance
    let final_balance = *account_balance.lock().unwrap();
    println!("Final account balance: {}", final_balance);
}

File: ga/gpt3_5_result_2024_6/result/multi_thread_basic_gpt/14_multi_thread_basic.rs
use std::sync::{Arc, Mutex};
use std::thread;

// Global shared variable representing account balance
static mut ACCOUNT_BALANCE: i32 = 1000000;

fn withdraw() {
    let amount = 500000;
    for _ in 0..amount {
        unsafe {
            let mutex = &MUTEX;
            let mut balance = mutex.lock().unwrap();
            *balance -= 1;
        }
    }
}

// Mutex for protecting the account balance
lazy_static::lazy_static! {
    static ref MUTEX: Arc<Mutex<i32>> = Arc::new(Mutex::new(ACCOUNT_BALANCE));
}

fn main() {
    let mut threads = vec![];

    let mtx = MUTEX.clone();
    threads.push(thread::spawn(move || {
        withdraw();
    }));

    let mtx2 = MUTEX.clone();
    threads.push(thread::spawn(move || {
        withdraw();
    }));

    for t in threads {
        t.join().unwrap();
    }

    unsafe {
        println!("Final account balance: {}", ACCOUNT_BALANCE);
    }
}